File: day5-csharp/Program.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

class Day5
{
    private const string RuleDivider = "|";
    private const string UpdatesDivider = ",";

    static void Main(string[] args)
    {
        var result = Part1("test-data");
        Console.WriteLine(result);
    }

    public static int Part1(string fileName = "test-data")
    {
        var (rules, updates) = ReadFile(fileName);
        var (preRules, postRules) = MapRules(rules);

        return updates
            .Select(MapUpdates)
            .Where(x => !CheckUpdate(x, preRules))
            .Select(x => ReorderElements(x, preRules, postRules))
            .Select(GetMiddleElement)
            .Sum();
    }

    private static (List<string>, List<string>) ReadFile(string fileName)
    {
        var lines = File.ReadAllLines($"./lib/{fileName}.txt").Select(line => line.Trim()).ToList();
        var splitIndex = lines.IndexOf("");
        return (lines.Take(splitIndex).ToList(), lines.Skip(splitIndex + 1).ToList());
    }

    private static (Dictionary<string, List<string>>, Dictionary<string, List<string>>) MapRules(List<string> rules)
    {
        var preRule = new List<(string, string)>();
        var postRule = new List<(string, string)>();

        foreach (var rule in rules)
        {
            var parts = rule.Split(RuleDivider);
            preRule.Add((parts[0], parts[1]));
            postRule.Add((parts[1], parts[0]));
        }

        return (CreateRuleMap(preRule), CreateRuleMap(postRule));
    }

    private static Dictionary<string, List<string>> CreateRuleMap(List<(string, string)> rules)
    {
        var map = new Dictionary<string, List<string>>();
        foreach (var (key, ruleItem) in rules)
        {
            if (!map.ContainsKey(key))
            {
                map[key] = new List<string>();
            }
            map[key].Add(ruleItem);
        }
        return map;
    }

    private static List<string> MapUpdates(string update)
    {
        return update.Split(UpdatesDivider).ToList();
    }

    private static bool CheckUpdate(List<string> update, Dictionary<string, List<string>> preRules)
    {
        var updateCount = update.Count;
        var indexes = from pre in Enumerable.Range(0, updateCount)
                      from post in Enumerable.Range(0, updateCount)
                      where pre < post
                      select (update[pre], update[post]);

        foreach (var (pre, post) in indexes)
        {
            if (preRules.ContainsKey(post))
            {
                var pres = preRules[post];
                if (pres.Contains(pre))
                {
                    return false;
                }
            }
        }
        return true;
    }

    public static List<string> ReorderElements(List<string> update, Dictionary<string, List<string>> preRules, Dictionary<string, List<string>> postRules)
    {
        return update.OrderBy(a =>
        {
            if (preRules.ContainsKey(a))
            {
                return preRules[a].Contains(a) ? -1 : 1;
            }
            if (postRules.ContainsKey(a))
            {
                return postRules[a].Contains(a) ? 1 : -1;
            }
            return 0;
        }).ToList();
    }

    private static int GetMiddleElement(List<string> update)
    {
        return int.Parse(update[update.Count / 2]);
    }
}

File: day5-csharp/program.zig
const std = @import("std");

const RuleMap = std.HashMap([]const u8, []const u8, std.hash_map.StringHashFn);

fn readFile(file_name: []const u8) !std.ArrayList([]const u8) {
    var file = try std.fs.cwd().openFile(file_name, .{});
    defer file.close();

    var allocator = std.heap.page_allocator;
    var lines = std.ArrayList([]const u8).init(allocator);

    while (true) {
        var line = try file.reader().readUntilDelimiterOrEofAlloc(allocator, '\n');
        if (line.len == 0) break;
        try lines.append(line);
    }

    return lines;
}

fn createRuleMap(rules: [][]const u8) RuleMap {
    var allocator = std.heap.page_allocator;
    var map = RuleMap.init(allocator);

    for (rules) |rule| {
        const parts = std.mem.split(rule, "|");
        if (parts.len == 2) {
            map.put(parts[0], parts[1]);
        }
    }

    return map;
}

fn mapUpdates(updates: [][]const u8) [][][]const u8 {
    var allocator = std.heap.page_allocator;
    var result = std.ArrayList([][][]const u8).init(allocator);

    for (updates) |update| {
        if (update.len != 0) {
            var parts = std.mem.split(update, ",");
            try result.append(parts);
        }
    }

    return result.toOwnedSlice();
}

fn checkUpdate(update: [][]const u8, preRules: RuleMap) bool {
    var update_count = update.len;

    for (0..update_count) |pre| {
        for (pre+1..update_count) |post| {
            if (preRules.get(update[post]) == update[pre]) {
                return false;
            }
        }
    }

    return true;
}

fn reorderElements(update: [][]const u8, preRules: RuleMap, postRules: RuleMap) [][]const u8 {
    std.sort.sort(update, std.sort.comparator([]const u8, a, b) {
        if (preRules.get(a) == b) return -1;
        if (postRules.get(b) == a) return 1;
        return 0;
    });

    return update;
}

fn getMiddleElement(update: [][]const u8) i32 {
    return std.fmt.parseInt(i32, update[update.len / 2], 10);
}

pub fn main() !void {
    const file_name = "test-data.txt";
    const rules_updates = try readFile(file_name);
    const rules = rules_updates[0..rules_updates.len / 2];
    const updates = rules_updates[rules_updates.len / 2..];

    const preRules = createRuleMap(rules);
    const postRules = createRuleMap(rules);

    var result = 0;

    for (mapUpdates(updates)) |update| {
        if (!checkUpdate(update, preRules)) {
            result += getMiddleElement(reorderElements(update, preRules, postRules));
        }
    }

    std.debug.print("Result: {}\n", .{result});
}